### PR TITLE
Loadout Editor v3

### PR DESCRIPTION
**Pull request by bhollis**
I'm still working on this, but I'd be happy to have folks kick the tires. The major missing piece is that there's still no way to swap an item from being equipped to not-equipped. There's more in TODOs, I'll start tracking them here. I also have a lot of code cleanup to do, and I'll peel some of the changes out into their own PRs.

Beyond that it's a pretty major redesign/rework of how loadouts work! I'll list out all the changes later, let me know if anything's confusing.

Screen Shot 2022-02-07 at 11 07 23 PM

Followups are at https://github.com/DestinyItemManager/DIM/issues/7841

### CODE REVIEW

**Comment by bhollis** [comment]
I spent a long while trying to get a drag interaction working, and it was OK:

Screen Shot 2022-02-09 at 11 48 09 PM

But it didn't work out for two reasons:

1. For some reason the buttons/items underneath the drop targets prevented dropping, so you could only drop on the spaces between items!
2. You couldn't drag items that were already in loadouts, because something about the drag targets appearing over the item cancels the drag.

**Comment by bhollis** [comment]
I'd like to get this into Beta only to start getting feedback - even if I have stuff I still want to do. If folks want to take a look at the code, that's cool - it's really a mess.

**Comment by nev-r** [comment]
image

**Comment by nev-r** [comment]
- the effects of **General->Fill from inventory** were unexpected and overwhelming
- `+ Add equipped` / `Add unequipped` could use rewording to maybe include the word "all" or something..
- love the `⋮` menus
- really appreciate the ability to remove optimizer settings
- maybe "fill from unequipped character inventory" or something... hm.
- would prefer "Save Changes" to "Update"
- notes setup is not ideal the way it takes so much room even if collapsed.. maybe a button to the right of title?

**Comment by nev-r** [comment]
> What would you have expected?

less "wrong thing happened", more "immediately added 18 tiles of height to the loadout sheet."

know what would be excellent, is if we added a (36) here
image
as a warning of what it would be adding

**Comment by nev-r** [comment]
- bungie.net calls them "non-equipped" and i agree it's a better and correcter adjective
- "fill" is definitely the right verb, "add" would mean include existing LO items and add all others as well, "use" would mean replace existing.. but "Fill in" would be more accurate or at least more specific. you "fill" a bucket of water but you "fill in" a set of boxes or fields.
- "from" is an unclear preposition, i think i would prefer "using"
- "using" unlocks describing nouns instead of a location, so i think my final result is this, but. maybe it's wordy..

"Fill in using equipped items"
"Fill in using non-equipped items"

and again, the numeric parenthetical reminder would help clarify/reassure that you're not saying "my entire vault"

File: src/app/loadout-drawer/LoadoutDrawerFooter.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+  const saveAsNewDisabled =
+    saveDisabled ||
+    // There's an existing loadout with the same name & class
+    Boolean(clashingLoadout);
+
+  return (
+    <div className={styles.loadoutOptions}>
+      {clashingLoadoutWarning && <div>{clashingLoadoutWarning}</div>}
+      <form onSubmit={onSaveLoadout}>
```
Comment (robojumper) [inline src/app/loadout-drawer/LoadoutDrawerFooter.tsx:79]:
```suggestion
      <form onSubmit={(e) => onSaveLoadout(e, isNew)}>
```

**Comment by robojumper** [comment]
Some things where people might see their existing workflows disrupted:

* Double click to toggle equipped doesn't have great discoverability (you acknowledged this already, just agreeing here)
* On mobile this also turns into an unreliable *triple* tap (about 50/50 between successfully toggling and bringing up item popup)! Item popup action button replacing "+ Loadout" button to toggle?
* When editing a loadout from loadouts page, only way to add arbitrary items to the non-equipped part of the loadout for a slot with an equipped item is to toggle that item first, then click the empty slot, select an item, then toggle the old item back in, which is a very roundabout way of doing it

**Comment by bhollis** [comment]
> When editing a loadout from loadouts page, only way to add arbitrary items to the non-equipped part of the loadout for a slot with an equipped item is to toggle that item first, then click the empty slot, select an item, then toggle the old item back in, which is a very roundabout way of doing it

Yeah - I'm not sure how much to encourage having non-equipped items in loadouts. One idea is to have extra little buttons for them that appear only if there's an equipped item. But you can also drag items in directly.

**Pull request merged**